File: frontend/src/types/gamification.ts
/**
 * Gamification types — mirror backend/app/schemas/gamification.py.
 */

export interface GamificationStats {
  total_xp: number;
  level: number;
  xp_into_level: number;
  xp_for_next_level: number;
  current_streak: number;
  longest_streak: number;
}

export interface Badge {
  code: string;
  label: string;
  description: string;
  icon: string; // Lucide icon name
  unlocked_at: string | null;
}

export interface RecordAttemptResult {
  xp_earned: number;
  leveled_up: boolean;
  new_level: number;
  stats: GamificationStats;
  newly_unlocked: Badge[];
}

export interface HistoryItem {
  quiz_id: string;
  score: number;
  understanding_level: string;
  xp_earned: number;
  completed_at: string;
  topic?: string | null;
}

export interface HistorySummary {
  total_quizzes: number;
  average_score: number;
  total_xp: number;
  best_score: number;
  worst_score: number;
  most_recent_topic?: string | null;
}

export interface TrendPoint {
  date: string;
  average_score: number;
  attempt_count: number;
}

export interface TopicMasteryItem {
  topic: string;
  average_score: number;
  attempt_count: number;
}

export interface GamificationAnalytics {
  quiz_count: number;
  average_score: number;
  total_xp: number;
  score_trend: TrendPoint[];
  topic_mastery: TopicMasteryItem[];
}

export interface HistoryResponse {
  summary: HistorySummary;
  items: HistoryItem[];
}

// §4.8 Batch 2 — Leaderboard + Weekly goal
export interface LeaderboardEntry {
  rank: number;
  name: string;
  total_xp: number;
  level: number;
  is_you: boolean;
}

export interface LeaderboardResponse {
  entries: LeaderboardEntry[];
  you_rank: number | null;
}

export interface WeeklyProgress {
  completed: number;
  target: number;
  percent: number;
  goal_met: boolean;
  remaining: number;
}

// §4.8 Batch 2-B — Preferences + Material bookmarks
export interface UserPreferences {
  default_num_questions: number;
  default_difficulty: "easy" | "medium" | "hard";
  shuffle_options: boolean;
  weekly_goal: number;
  favorite_topic: string | null;
}

export interface Bookmark {
  id: string;
  title: string;
  material_text: string;
  created_at: string;
}

export interface BookmarkList {
  items: Bookmark[];
}

File: backend/app/schemas/gamification.py
"""Public schemas for gamification endpoints.

Any change here must be mirrored in frontend/src/types/gamification.ts.
"""

from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field


class RecordAttemptRequest(BaseModel):
    """POST /gamification/record-attempt body."""

    quiz_id: str = Field(..., min_length=1, max_length=64)
    score: int = Field(..., ge=0, le=100)
    understanding_level: str = Field(..., min_length=1, max_length=16)


class StatsResponse(BaseModel):
    total_xp: int
    level: int
    xp_into_level: int
    xp_for_next_level: int
    current_streak: int
    longest_streak: int


class BadgeResponse(BaseModel):
    code: str
    label: str
    description: str
    icon: str
    unlocked_at: datetime | None = None


class RecordAttemptResponse(BaseModel):
    xp_earned: int
    daily_bonus_earned: int = 0
    leveled_up: bool
    new_level: int
    stats: StatsResponse
    newly_unlocked: list[BadgeResponse]


class HistoryItem(BaseModel):
    quiz_id: str
    score: int
    understanding_level: str
    xp_earned: int
    completed_at: datetime
    topic: str | None = None


class HistorySummary(BaseModel):
    total_quizzes: int
    average_score: int
    total_xp: int
    best_score: int
    worst_score: int
    most_recent_topic: str | None = None


class HistoryResponse(BaseModel):
    summary: HistorySummary
    items: list[HistoryItem]


class TrendPoint(BaseModel):
    date: date
    average_score: int
    attempt_count: int


class TopicMasteryItem(BaseModel):
    topic: str
    average_score: int
    attempt_count: int


class AnalyticsResponse(BaseModel):
    quiz_count: int
    average_score: int
    total_xp: int
    score_trend: list[TrendPoint]
    topic_mastery: list[TopicMasteryItem]


# ============================================================================
# §4.8 Batch 2 — Leaderboard + Weekly goal
# ============================================================================


class LeaderboardEntry(BaseModel):
    rank: int
    name: str
    total_xp: int
    level: int
    is_you: bool = False


class LeaderboardResponse(BaseModel):
    entries: list[LeaderboardEntry]
    # The requester's rank even when they're outside the returned top-N
    # (None when they have no recorded progress yet).
    you_rank: int | None = None


class WeeklyProgressResponse(BaseModel):
    completed: int
    target: int
    percent: int
    goal_met: bool
    remaining: int


# ============================================================================
# §4.8 Batch 2-B — Preferences + Material bookmarks
# ============================================================================

from typing import Literal, Optional


class PreferencesResponse(BaseModel):
    default_num_questions: int
    default_difficulty: str
    shuffle_options: bool
    weekly_goal: int
    favorite_topic: Optional[str] = None


class PreferencesUpdate(BaseModel):
    """All fields optional — a PATCH-style partial update."""

    default_num_questions: Optional[Literal[3, 5, 7, 10]] = None
    default_difficulty: Optional[Literal["easy", "medium", "hard"]] = None
    shuffle_options: Optional[bool] = None
    weekly_goal: Optional[int] = Field(default=None, ge=1, le=50)
    favorite_topic: Optional[str] = Field(default=None, max_length=80)


class BookmarkCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=120)
    material_text: str = Field(..., min_length=1, max_length=20_000)


class BookmarkItem(BaseModel):
    id: str
    title: str
    material_text: str
    created_at: datetime


class BookmarkListResponse(BaseModel):
    items: list[BookmarkItem]
